File: ALGORITMS/7. GREEDY ALGORITHM/leetcode-remove-k-digits.py
# url :- https://leetcode.com/problems/remove-k-digits/

from collections import deque


class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        if k >= len(num) :
            return "0"
        
        if len(num) <= 1:
            return num
        
        n = len(num)
        stack = []
    
        for ch in num:
            while k > 0 and stack and stack[-1] > ch:
                k -= 1
                stack.pop()
            
            stack.append(ch)

        stack = stack[:len(stack) - k]
        res = ""
        last = True 
        for ch in stack :
            if ch == '0' and last :
                continue
            if ch != '0' :
                last = False 
            res += ch 
        return res

    
sol = Solution()
print(sol.removeKdigits("1432219", 3) == '1219')
print(sol.removeKdigits("10200", 1) == '200')
print(sol.removeKdigits("10", 2) == "0")   

File: ALGORITMS/7. GREEDY ALGORITHM/leetcode-remove-k-digits.ts
function removeKdigits(num: string, k: number): string {
	const stack: string[] = [];
	let hashMap = new Set();

	for (const ch of num) {
		if (k > 0 && stack.length > 0 && parseInt(stack[stack.length - 1]) > parseInt(ch)) {
			k--;
			stack.pop();
		}
		stack.push(ch);
	}
	const res = stack.splice(0, stack.length - k);
	let ans = "";
	let last = true;
	for (const ch of res) {
		if (ch == "0" && last) {
			continue;
		}
		if (ch != "0") last = false;
		ans += ch;
	}
	return ans == "" ? "0" : ans;
}
